server/plugin.go

```diff
@@ -138,17 +139,44 @@ func (p *Plugin) GetURL() string {
 	return *config.ServiceSettings.SiteURL + "/plugins/" + pluginID
 }
 
+func (p *Plugin) OnDisconnectedTokenHandler(userID string) {
+	p.API.LogDebug("OnDisconnectedTokenHandler", "userID", userID)
+	teamsUserID, err := p.store.MattermostToTeamsUserID(userID)
+	if err != nil {
+		p.API.LogWarn("Unable to get teams user id from mattermost to user", "userID", userID, "error", err.Error())
+		return
+	}
+	if err2 := p.store.SetUserInfo(userID, teamsUserID, nil); err2 != nil {
+		p.API.LogWarn("Unable clean invalid token for the user", "userID", userID, "error", err2.Error())
+		return
+	}
+	channel, appErr := p.API.GetDirectChannel(userID, p.GetBotUserID())
+	if appErr != nil {
+		p.API.LogWarn("Unable to get direct channel for send message to user", "userID", userID, "error", appErr.Error())
+		return
+	}
+	_, appErr = p.API.CreatePost(&model.Post{
+		UserId:    p.GetBotUserID(),
+		ChannelId: channel.Id,
+		Message:   "Your connection to MS Teams has been lost. Please reconnect to using '/msteams-sync connect' slash command.",
+	})
+	if appErr != nil {
+		p.API.LogWarn("Unable to send direct message to user", "userID", userID, "error", appErr.Error())
+	}
+}
```

Comment:
I'm hesitant to /immediately/ send a message to the enduser. There's a chance that events come into multiple plugins (or even activity handlers) and we fire off a rash of messages to the same user.

One approach might be to introduce a new job whose responsibility it is to notify these users, and add a column `DisconnectNotifiedAt` to track when last done. I'm leaning towards being super robust here, especially if there's a case where we "miss" sending a message -- the job would allow us to seamlessly catch up and notify users as such.

Reply:
Fair enough, I like the idea of having it persisted and ensure it is sent.